### ISSUE TITLE
System.OverflowException when trying to read string value from an array

### ISSUE THREAD

**Issue by EranOfer** [microsoft/clrmd#244]
Hi ,
I getting a System.OverflowException when trying to read string value from an array.
As you can see it throw OverflowException after reading 30 item with no problem.

I working on Microsoft.Diagnostics.Runtime version="1.0.2" .
The dump is from wpf app 4.5.1 with 64 bit.
 
![image](https://user-images.githubusercontent.com/13414515/54884922-3ea00b80-4e7f-11e9-95c4-986b4a9bad69.png)

My code:
```c#
ulong address = 2697244618824;
               var type = clrHeap.GetObjectType(address);// clrObject.Type;
                if (type.IsArray && (type.ComponentType.IsPrimitive || type.ComponentType.IsString))
                {
                    var arrayLength = type.GetArrayLength(address);
                    for (int i = 0; i < arrayLength; i++)
                    {
                        var key = type.ComponentType.GetValue(type.GetArrayElementAddress(address, i))
                            .ToString();

                    }
                }
```

Thanks in advance for your help,
Eran

**Comment by mitikov** [comment]
Hey @EranOfer, does it happen constantly or once in a while/for single string?

The string layout in memory would be smth like : `typeId`_`Length`_`charactersGoHere`

I recall seeing similar trouble once in a while when `Length` gets some trash value (zillions).

What is the outcome if you open a snapshot in WinDbg and call `!do` command?

> 0:079> !do 00000291b2c20fb8
> Name:        System.String
> MethodTable: 00007ff9c0149798
> EEClass:     00007ff9bfa26cb8
> Size:        76(0x4c) bytes
> File:        C:\Windows\Microsoft.Net\assembly\GAC_64\mscorlib\v4.0_4.0.0.0__b77a5c561934e089\mscorlib.dll
> String:      api/v1/Imports/ImportData
> Fields:
>               MT    Field   Offset                 Type VT     Attr            Value Name
> **00007ff9c014c0e8  400027b        8         System.Int32  1 instance               25 m_stringLength**
> 00007ff9c014a950  400027c        c          System.Char  1 instance               61 m_firstChar
> 00007ff9c0149798  4000280       c0        System.String  0   shared           static Empty
>                                  >> Domain:Value  00000290f08a0d60:NotInit  00000294ccae0630:NotInit  <<

Or: **dp 00000291b2c20fb8**
00000291b2c20fb8  00007ff9c0149798 00700061`000000**19**  (hex 19 = 25 dec)

The solution I took last time was to make a [ReadStringSafe](https://github.com/mitikov/Sitecore.MemoryDiagnostics/blob/master/src/Sitecore.MemoryDiagnostics/PrimitiveEntitiesReaders/StringReader.cs#L24) method that does not read more than ~3k chars out of the box.

Thanks.

**Comment by leculver** [comment]
There's not much we can do in general when there's a giant string or somehow the value is corrupted.  However, this is a great idea:

> The solution I took last time was to make a ReadStringSafe method that does not read more than ~3k chars out of the box.

I will add this to ClrMD 2.0.

**Comment by leculver** [comment]
This is fixed in clrmd 2.0.  The default is to cap strings at a maximum of 4096 characters, but these methods will allow you to specify int.MaxValue to uncap the strings.
